sklearn/utils/validation.py

```diff
@@ -80,7 +80,9 @@ def as_float_array(X, copy=True, force_all_finite=True):
     elif X.dtype in [np.float32, np.float64]:  # is numpy array
         return X.copy('F' if X.flags['F_CONTIGUOUS'] else 'C') if copy else X
     else:
-        return X.astype(np.float32 if X.dtype == np.int32 else np.float64)
+        return X.astype(np.float32 if X.dtype.kind in 'uib' and \
```

Comment:
Rather than cramming too much in a single statement, I would do something like:

```py
if X.dtype.kind in 'uib' and X.dtype.itemsize <= 4:
    return_dtype = np.float32
else:
    return_dtype = np.float64
return X.astype(return_dtype)
```